File: 347.前-k-个高频元素.cpp
/*
 * @lc app=leetcode.cn id=347 lang=cpp
 *
 * [347] 前 K 个高频元素
 */
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

// @lc code=start
class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        unordered_map<int, int> record;
        for (int i = 0; i < nums.size(); i++)
        {
            record[nums[i]]++;
        }
        int n = record.size();
        // 小根堆
        priority_queue<pair<int, int>,
                       vector<pair<int, int> >,
                       greater<pair<int, int> > >
            pq;
        for (auto it = record.begin(); it != record.end(); it++)
        {
            if (k == pq.size())
            {
                if (it->second > pq.top().first)
                {
                    pq.pop();
                    pq.push(make_pair(it->second, it->first));
                }
            }
            else
            {
                pq.push(make_pair(it->second, it->first));
            }
        }

        vector<int> result;
        while (!pq.empty())
        {
            result.push_back(pq.top().second);
            pq.pop();
        }
        reverse(result.begin(), result.end());
        return result;
    }
};
// @lc code=end

File: 347.前-k-个高频元素.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

/*
 * @lc app=leetcode.cn id=347 lang=java
 *
 * [347] 前 K 个高频元素
 */

// @lc code=start
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        // 适用map记录
        for (int num : nums) {
            map.putIfAbsent(num, 0);
            map.put(num, map.get(num) + 1);
        }

        // 遍历map， 用最小堆保存最大的k个元素
        PriorityQueue<Integer> pq = new PriorityQueue<>(new Comparator<Integer>() {

            @Override
            public int compare(Integer o1, Integer o2) {
                return map.get(o1) - map.get(o2);
            }
        });

        for (Integer key : map.keySet()) {
            if (pq.size() < k) {
                pq.add(key);
            } else if (map.get(key) > map.get(pq.peek())) {
                pq.remove();
                pq.add(key);
            }
        }
        int[] res = new int[pq.size()];
        int i = 0;
        while (!pq.isEmpty()) {
            res[i++] = pq.remove();
        }
        return res;
    }
}
// @lc code=end
